README.md
![Logo](static/unisystemrfq.png)

# `Backend` in Node.js & Typescript for **Riverdi RFQ** System

**Notice: This project is a REST API that is ment to work with [frontend GUI](link)**

<hr>

> The sole purpose of this project is to fullfill requirements of **[Riverdi](https://riverdi.com)** stakeholders (Riverdi Sales Department) in terms of creating an application for convienient **RFQ** (_Request For Quotation_) mangement and integrate it with current workflows that are run in [Comarch ERP XL](https://www.comarch.pl/erp/xl/), [ClickUp](https://clickup.com/) and [SharePoint](https://www.microsoft.com/pl-pl/microsoft-365/sharepoint/collaboration).

<hr>

## Main Features

- Add new RFQs
- Generate unique RFQ code (based on specification)
- list RFQs
- sort RFQs (handled on frontend)
- filter RFQs (handled on frontend)
- Add requirements (with notes) to existing RFQs
- All `CRUD` operations on: **RFQs, requirements, distributors and users**
- Comarch ERP XL integration (fetching customers)
- SharePoint Integration (creating specified folder structure, puting files)
- ClickUp Integration (creating tasks, assigning people, fetching task statuses)

## Data Model

![Data Model](static/riverdiRfqDataModel.png)

## Migrations

In order to tie the database structure to the code (helpful when working with other engineers) I had to use **`schema migrations`** (_code that describes a precise change to make to the database_).

The tool of the choice is [node-pg-migrate](https://github.com/salsita/node-pg-migrate). All migratrons are written in plain SQL.

> What are **`schema migrations`** and how to use **node-pg-migrate** is out of the scope of this document.

## Data Layer Logic

I've created separate layer of utility methods to handle database operations.

They should be used in controller business logic instead of operating directly on database.

---

### User Repo

---

#### Method:

```Javascript
UserRepo.find();
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, username, name, email, shortname, role_id}]
```

> Returns list of all users that are not admins and are not marked as `deleted`.

---

#### Method:

```Javascript
UserRepo.findWithAdmins();
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, username, name, email, shortname, role_id}]
```

> Returns list of all users (admins included) that are not marked as `deleted`.

---

#### Method:

```Javascript
UserRepo.findById(id: string);
```

#### Response (`Object`):

```Javascript
{id, username, email, shortname, role_id}
```

> Returns user with matching `id` (or returns empty object).

---

#### Method:

```Javascript
UserRepo.findByEmail(email: string);
```

#### Response (`Object`):

```Javascript
{id, username, email, password, shortname, role_id, deleted}
```

> Returns user with matching email (or returns empty object). Returned **hashed** password is for login logic purposes.

---

#### Method:

```Javascript
UserRepo.insert({
  username,
  password,
  email,
  shortname,
  role_id,
}: {
  username: string;
  password: string;
  email: string;
  shortname: string;
  role_id: string;
});
```

#### Response (`Object`):

```Javascript
{id, username, email, shortname, role_id}
```

> Inserts new user into database. Shortname and email have to be unique. **Password should be stored in a hashed form!** _(route controller logic should handle that)_

---

#### Method:

```Javascript
UserRepo.updateData({
  id,
  username,
  email,
  shortname,
  role_id,
}: {
  id: string;
  username: string;
  email: string;
  shortname: string;
  role_id: string;
});
```

#### Response (`Object`):

```Javascript
{id, username, email, shortname, role_id}
```

> Updates existing user's data (only username, email, shortname and role_id). Shortname and email have to be unique. ID is immutable. Changing password is handled by dedicated method.

---

#### Method:

```Javascript
UserRepo.updatePassword({
  id,
  password,
}: {
  id: string;
  password: string;
});
```

#### Response (`Object`):

```Javascript
{id, username, email, shortname, role_id}
```

> Updates existing user's password. **Password should be provided in a hashed form!** _(route controller logic should handle that)_

---

#### Method:

```Javascript
UserRepo.delete(id: string);
```

#### Response (`Object`):

```Javascript
{id, username, email, role_id, shortname}
```

> Removes user from database.

---

#### Method:

```Javascript
UserRepo.markDeleted(id: string)
```

#### Response (`Object`):

```Javascript
{id, username, email, role_id, shortname}
```

> Marks user ad "deleted". Disables user (from login, from preforming actions, from appearing on all kings of listings) but keeps all the data (transaction history, etc.)

---

#### Method:

```Javascript
UserRepo.markUndeleted(id: string)
```

#### Response (`Object`):

```Javascript
{id, username, email, role_id, shortname}
```

> Sets "deleted" flag to `false` (enables user).

---

#### Method:

```Javascript
UserRepo.count()
```

#### Response (`Number`):

> Returns number of users.

---

### RFQ Repo

---

#### Method:

```Javascript
RfqRepo.find();
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, rfq_code, eau, customer, distributor, pm, kam, updated}]
```

> Returns list of all rfqs

---

#### Method:

```Javascript
RfqRepo.findById(id: string);
```

#### Response (`Object`):

```Javascript
{
  id,
  rfq_code,
  eau,
  clickup_id,
  customer_id,
  customer,
  distributor_id,
  distributor,
  pm_id,
  pm,
  pm_fullname,
  kam_id,
  kam,
  kam_fullname,
  final_solutions,
  conclusions,
  samples_expected,
  mp_expected,
  eau_max,
  updated
}
```

> Returns RFQ with matching `id`.

---

#### Method:

```Javascript
RfqRepo.findByDistributorId(distributor_id: string)
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, rfq_code}]
```

> Returns list of all rfqs of distributor with matching `distributor_id`.

---

#### Method:

```Javascript
RfqRepo.findByRfqCode(rfq_code: string);
```

#### Response (`Object`):

```Javascript
{id, rfq_code}
```

> Returns RFQ with matching `id`.

---

#### Method:

```Javascript
RfqRepo.insert({
   rfq_code,
    eau,
    customer_id,
    distributor_id,
    pm_id,
    kam_id,
    clickup_id,
    final_solutions,
    conclusions,
    samples_expected,
    mp_expected,
    eau_max,
  }: {
    rfq_code: string;
    eau: string;
    customer_id: string;
    distributor_id: string;
    pm_id: string;
    kam_id: string;
    clickup_id: string;
    final_solutions: string;
    conclusions: string;
    samples_expected: string;
    mp_expected: string;
    eau_max: string;
});
```

#### Response (`Object`):

```Javascript
{id, rfq_code}
```

> Inserts new RFQ into database. `rfq_code` have to be unique.

---

#### Method:

```Javascript
RfqRepo.updateData({
  id,
    eau,
    customer_id,
    distributor_id,
    pm_id,
    kam_id,
    final_solutions,
    conclusions,
    samples_expected,
    mp_expected,
    eau_max,
  }: {
    id: string;
    eau: string;
    customer_id: string;
    distributor_id: string;
    pm_id: string;
    kam_id: string;
    final_solutions: string;
    conclusions: string;
    samples_expected: string;
    mp_expected: string;
    eau_max: string;
});
```

#### Response (`Object`):

```Javascript
{id, rfq_code}
```

> Updates RFQ into database. `rfq_code` have to be unique.

---

#### Method:

```Javascript
RfqRepo.delete(id: string);
```

#### Response (`Object`):

```Javascript
{id, rfq_code}
```

> Removes rfq from database.

---

#### Method:

```Javascript
RfqRepo.count()
```

#### Response (`Number`):

> Returns number of rfqs.

---

### Requirement Repo

---

#### Method:

```Javascript
RequirementRepo.find();
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, rfq_id, priority, c_nc_cwr, requirement, note, date }]
```

> Returns list of all requirements

---

#### Method:

```Javascript
RequirementRepo.findByRfqId(rfq_id: string);
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, rfq_id, priority, c_nc_cwr, requirement, note, date }]
```

> Returns list of all requirements for given `rfq_id`

---

#### Method:

```Javascript
RequirementRepo.findById(id: string);
```

#### Response (`Object`):

```Javascript
{id, rfq_id, priority, c_nc_cwr, requirement, note, date }
```

> Returns requirement with matching `id`

---

#### Method:

```Javascript
RequirementRepo.insert({
    rfq_id,
    priority,
    c_nc_cwr,
    requirement,
    note,
  }: {
    rfq_id: string;
    priority: string;
    c_nc_cwr: string;
    requirement: string;
    note: string;
});
```

#### Response (`Object`):

```Javascript
{id, rfq_id}
```

> Inserts new **requirement** into database.

---

#### Method:

```Javascript
RequirementRepo.updateData({
    id,
    rfq_id,
    priority,
    c_nc_cwr,
    requirement,
    note,
  }: {
    id: string;
    rfq_id: string;
    priority: string;
    c_nc_cwr: string;
    requirement: string;
    note: string;
});
```

#### Response (`Object`):

```Javascript
{id, rfq_id, c_nc_cwr, requirement, note}
```

> Updates **requirement** with given `id` in the database.

---

#### Method:

```Javascript
RequirementRepo.delete(id: string);
```

#### Response (`Object`):

```Javascript
{id}
```

> Removes **requirement** from database.

---

#### Method:

```Javascript
RequirementRepo.count()
```

#### Response (`Number`):

> Returns number of requirements.

---

### Distributor Repo

---

#### Method:

```Javascript
DistributorRepo.find();
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, name }]
```

> Returns list of all distributors

---

#### Method:

```Javascript
DistributorRepo.findById(id: string);
```

#### Response (`Object`):

```Javascript
{id, name }
```

> Returns distributor with matching `id`

---

#### Method:

```Javascript
DistributorRepo.findByName(name: string);
```

#### Response (`Object`):

```Javascript
{id, name }
```

> Returns distributor with matching `name`

---

#### Method:

```Javascript
DistributorRepo.insert({ name }: { name: string });
```

#### Response (`Object`):

```Javascript
{id, name}
```

> Inserts new distributor into database.

---

#### Method:

```Javascript
DistributorRepo.updateData({ id, name }: { id: string; name: string });
```

#### Response (`Object`):

```Javascript
{id, name}
```

> Updates distributor with given `id` in the database.

---

#### Method:

```Javascript
DistributorRepo.delete(id: string);
```

#### Response (`Object`):

```Javascript
{id, name}
```

> Removes distributor from database.

---

#### Method:

```Javascript
DistributorRepo.count()
```

#### Response (`Number`):

> Returns number of distributors.

---

### Customer Repo

---

#### Method:

```Javascript
CustomerRepo.find();
```

#### Response (`Array` of `Objects`):

```Javascript
[{id, name }]
```

> Returns list of all customers

---

#### Method:

```Javascript
CustomerRepo.findById(id: string);
```

#### Response (`Object`):

```Javascript
{id, name }
```

> Returns customer with matching `id`

---

#### Method:

```Javascript
CustomerRepo.count()
```

#### Response (`Number`):

> Returns number of customers.

---

## API (endpoints documentation)

---

### Usage _(note: this is a REST API and it is supposed to be consumed by some kind of frontend)_

---

### 👉 User Router:

---

#### Route (unprotected): `/api/v1/users/login`

> Request: `POST`
>
> Request Body: `{email, password}`
>
> Response: `{ id, username, email, shortname, role_id, deleted }` + `Set-Cookie` with **user** in **`JWT`**

_**Feature:** allows existing user to "log in" - sends back user details in response body, starts cookie session and sends back user in JWT ([JSON Web Token](https://jwt.io/)) stored in a cookie_

---

#### Route (protected): `/api/v1/users/signup`

> Request: `POST`
>
> Request Body: `{ email, password, passwordConfirm, username, shortname, role_id }`
>
> Request Cookie: `currentUser in JWT`
>
> Response: `{ id, username, email, shortname, role_id }`

_**Feature:** create a new user with data provided in the request's body (important! New users can be added only by admins – standard user would get 401 error with message: **"Not authorized"** )_

---

#### Route (unprotected): `/api/v1/users/logout`

> Request: `POST`
>
> Request Body: `{ email, password, passwordConfirm, username, shortname, role_id }`
>
> Response: `{ id, username, email, shortname, role_id }`

_**Feature:** logs out user - clears session_

---

#### Route (protected): `/api/v1/users/:id"`

> Request: `GET`
>
> Request Cookie: `currentUser in JWT`
>
> Response: `{ id, username, email, shortname, role_id }`

_**Feature:** returns user's `id, username, email, shortname, role_id` for given `id`_

---

## Tests

## Screenshots

## How to use it

## How to develop it

git flow

## Future optimalisations

server side sorting + data pagination

## Infrastructure (backend, frontend, database, cookies)

## Security

## Installation

## Configuration

# Project Name

> Abstract of the project or small introduction of what the project is about

<hr>

## Table of Contents

- [Team Members](#team-members)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.9.0] - 2022-06-18

### 🔥 Added

- offers module

## [2.8.0] - 2022-06-05

### 🔥 Added

- started offers module

## [2.7.0] - 2022-05-19

### 🔥 Added

- five new requirements

## [2.6.0] - 2022-05-17

### 🔥 Added

- user access restrictions

## [2.5.0] - 2022-05-11

### 🔥 Added

- RFQ's 2.0

## [2.4.0] - 2022-04-28

### 🔥 Added

- `rndTasks` everything

## [2.3.0] - 2022-04-20

### 🔥 Added

- `project` ClickUp integration
- `project` SharePoint integration

## [2.2.0] - 2022-03-27

### 🔥 Added

- `project` repo CRUD
- `project` controller & router REST API CRUD
- `partnumber` repo CRUD
- `partnumber` controller & router REST API CRUD

## [2.1.0] - 2022-03-24

### 🔥 Added

- `industry` repo CRUD
- `industry` controller & router REST API CRUD

## [2.0.0] - 2022-03-23

### 🔥 Added

- `project-client` repo CRUD
- `projectClient` controller & router REST API CRUD

## [2.0.0] - 2022-03-18

### 🔥 Added

- converted into Unisystem

## [1.8.0] - 2022-03-17

### 💪 Updated

- JWT based authentication

## [1.8.1] - 2021-08-17

### 👾 Fixed

- RfqRepo.find() returns empty string instead of null in extra_note

## [1.8.0] - 2021-08-16

### 🔥 Added

- add `extra_note` text field to `rfqs` table via **db schema migrations**
- add `extra_note` to the backend - updated `RfqRepo` class and `RFQ routes` (POST, PUT, GET) controllers logic

## [1.7.0] - 2021-08-05

### 🔥 Added

- change `eau` -> `eau min` (mandatory) + `eau max` (optional) via **db schema migrations**
- add requirements `order` feature via **db schema migrations**
- add `final solutions` notes textarea and `conclusions` notes textarea via **db schema migrations**
- add `samples expected` and `mp expected` text fields via **db schema migrations**
- change `eau` -> `eau min` (mandatory) + `eau max` (optional) - updated `RfqRepo` class and `RFQ routes` (POST, PUT, GET) controllers logic
- add requirements `order` feature - updated `RequirementRepo` class and `Requirement routes` (POST, PUT, GET) - controllers logic
- autogenerating links to **Riverdi RFQ App** and **Sharepoint** and put them into **Clickup** task description - updated `RFQ routes` - controllers logic
- add `final solutions` notes textarea and `conclusions` notes textarea - updated `RfqRepo` class and `RFQ routes` - (POST, PUT, GET) controllers logic
- `extra_note` field to add after `rfq_code` in clickup task - updated `RFQ routes` controllers logic
- add `samples expected` and `mp expected` text fields - updated `RfqRepo` class and `RFQ routes` (POST, PUT, GET) - controllers logic
- unit tests for the new features

## [1.6.0] - 2021-08-02

### 🔥 Added

- schema migrations for new features

## [1.5.0] - 2021-07-10

### 🔥 Added

- test coverage for critical features

## [1.4.0] - 2021-07-06

### 💪 Updated

- guarded clickup task not found case

## [1.3.0] - 2021-07-05

### 🔥 Added

- ClickUp integration (`findUserId`, `createTask`, `getTaskStatus`)

### 💪 Updated

- RfqRepo
- `new RFQ` route logic
- `show RFQ` route logic

## [1.2.0] - 2021-07-01

### 🔥 Added

- added SharePoint integration

## [1.1.0] - 2021-06-29

### 🔥 Added

- disable `login` feature for "deleted" users
- logic for updating user data on PUT `/api/v1/users/:id` route
- logic for updating user's password on PUT `/api/v1/users/:id/changepassword` route
- logic for listing ALL users (admins included) for management purposes on GET `/api/v1/usersandadmins` route
- logic for marking users as deleted on POST `/api/v1/users/disable` route
- logic for marking users as undeleted on POST `/api/v1/users/enable` route

## [1.0.0] - 2021-06-27

### 🔥 Added

- logic for different fields and confirm password validation in `creating user`

## [0.9.1] - 2021-06-25

### 👾 Fixed

- format of returned `role_id`

## [0.9.0] - 2021-06-24

### 💪 Updated

- return additional user data with rfq
- return requirements sorted by creation timestamp
- return only users marked as not deleted
- users CRUD should be only possible for admins

## [0.8.0] - 2021-06-19

### 🔥 Added

- logic for creating new distributor on POST `/api/v1/distributors` route
- logic for listing all distributors on GET `/api/v1/distributors` route
- logic for updating requirement on PUT `/api/v1/requirements/:id` route
- logic for updating rfq on PUT `/api/v1/rfqs/:id` route
- logic for updating distributor on PUT `/api/v1/distributors/:id` route
- logic for deleting requirement on DELETE `/api/v1/requirements/:id` route
- logic for deleting rfq on DELETE `/api/v1/rfqs/:id` route
- logic for deleting distributor on DELETE `/api/v1/distributors/:id` route

## [0.7.0] - 2021-06-18

### 🔥 Added

- customerRepo db handler (orm style)
- distributorRepo db handler (orm style)
- logic for listing customers on GET `/api/v1/customers` route
- logic for listing distributors on GET `/api/v1/distributors` route

## [0.6.0] - 2021-06-16

### 💪 Updated

- changed userRepo and route to return all necessary user fields, sort records by username and filter out admins

## [0.5.0] - 2021-06-15

### 🔥 Added

- logic for showing a RFQ on GET `/api/v1/rfq/:id` route
- logic for showing a RFQ's requirements on GET `/api/v1/rfq/:id/requirements` route
- tests for showing a RFQ on GET `/api/v1/rfq/:id` route
- tests for showing a RFQ's requirements on GET `/api/v1/rfq/:id/requirements` route

## [0.4.0] - 2021-06-12

### 🔥 Added

- requirementRepo db handler (orm style)
- logic for creating new requirement on POST `/api/v1/requirements` route
- tests for creating new requirement on POST `/api/v1/requirements` route

## [0.3.0] - 2021-06-11

### 🔥 Added

- logic for listing rfqs on GET `/api/v1/rfqs` route
- tests for listing rfqs on GET `/api/v1/rfqs` route

## [0.2.0] - 2021-06-10

### 🔥 Added

- db schema for the app
- schema migrations

src/repos/rfq-repo.ts
import { BadRequestError } from "../errors";
import { pool } from "../pool";

class RfqRepo {
  static async find() {
    try {
      const result = await pool.query(`
      SELECT
        r.id,
        rfq_code,
        eau,
        department,
        for_valuation::VARCHAR,
        r.name AS extra_note,
        project_clients.name AS customer,
        kam.id AS kam_id,
        pm.shortname AS pm,
        kam.shortname AS kam,
        to_char(r.updated_at, 'YYYY-MM-DD HH24:MI:SS') as updated
        FROM rfqs AS r
        JOIN project_clients ON project_clients.id = r.project_client_id
        JOIN users AS pm ON pm.id = r.pm_id
        JOIN users AS kam ON kam.id = project_clients.kam_id
        WHERE r.id != 1
        ORDER BY updated DESC;
      `);
      return result?.rows;
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async findByKamId(kamId: string) {
    try {
      const result = await pool.query(
        `
      SELECT
        r.id,
        rfq_code,
        eau,
        department,
        for_valuation::VARCHAR,
        r.name AS extra_note,
        r.project_client_id AS project_client_id,
        project_clients.name AS customer,
        kam.id AS kam_id,
        pm.shortname AS pm,
        kam.shortname AS kam,
        to_char(r.updated_at, 'YYYY-MM-DD HH24:MI:SS') as updated
        FROM rfqs AS r
        JOIN project_clients ON project_clients.id = r.project_client_id
        JOIN users AS pm ON pm.id = r.pm_id
        JOIN users AS kam ON kam.id = project_clients.kam_id
        WHERE r.id != 1 AND kam.id = $1
        ORDER BY updated DESC;
      `,
        [kamId]
      );
      return result?.rows;
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async findById(id: string) {
    try {
      const result = await pool.query(
        `
        SELECT
              r.id,
              rfq_code,
              eau,
              r.department AS department,
              r.clickup_id AS clickup_id,
              r.project_client_id AS project_client_id,
              r.name AS name,
              sp,
              year,
              serial,
              project_clients.name AS customer,
              for_valuation,
              req_disp_tech,
              req_disp_size,
              req_disp_res,
              req_disp_brigt,
              req_disp_angle,
              req_disp_od,
              req_disp_aa,
              req_disp_inter,
              req_disp_ot,
              req_disp_st,
              req_disp_spec,
              req_tp_size,
              req_tp_aa,
              req_tp_tech,
              req_tp_od,
              req_tp_inter,
              req_tp_glass,
              req_tp_spec,
              req_others,
              r.pm_id AS pm_id,
              projects.id AS project_id,
              projects.project_code AS project_code,
              pm.shortname AS pm,
              pm.username AS pm_fullname,
              project_clients.kam_id AS kam_id,
              kam.shortname AS kam,
              kam.username AS kam_fullname,
              final_solutions,
              samples_expected,
              mp_expected,
              to_char(r.updated_at, 'YYYY-MM-DD HH24:MI:SS') as updated
              FROM rfqs AS r
              FULL OUTER JOIN projects ON projects.rfq_id = r.id
              JOIN project_clients ON project_clients.id = r.project_client_id
              JOIN users AS pm ON pm.id = r.pm_id
              JOIN users AS kam ON kam.id = project_clients.kam_id
        WHERE r.id = $1
        `,
        [id]
      );
      return result?.rows[0];
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async findByRfqCode(rfq_code: string) {
    try {
      const result = await pool.query(
        `SELECT id, rfq_code FROM rfqs WHERE rfq_code = $1;`,
        [rfq_code]
      );
      return result?.rows[0];
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async findMaxSerialForDeptAndYear({
    department,
    year,
  }: {
    department: string;
    year: number;
  }) {
    try {
      const result = await pool.query(
        `SELECT MAX(serial) FROM rfqs WHERE department = $1 AND year = $2;`,
        [department, year + ""]
      );

      console.log({ rfq_max_count: result?.rows[0] });
      const max = result?.rows[0].max;
      return max ? parseInt(max) : 0;
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async insert({
    rfq_code,
    clickup_id,
    sp,
    year,
    serial,
    department,
    eau,
    project_client_id,
    name,
    pm_id,
    samples_expected,
    mp_expected,
    for_valuation,
    req_disp_tech,
    req_disp_size,
    req_disp_res,
    req_disp_brigt,
    req_disp_angle,
    req_disp_od,
    req_disp_aa,
    req_disp_inter,
    req_disp_ot,
    req_disp_st,
    req_disp_spec,
    req_tp_size,
    req_tp_aa,
    req_tp_tech,
    req_tp_od,
    req_tp_inter,
    req_tp_glass,
    req_tp_spec,
    req_others,
  }: {
    rfq_code: string;
    clickup_id: string;
    sp: string;
    year: number;
    serial: number;
    department: string;
    eau: string;
    project_client_id: string;
    name: string;
    pm_id: string;
    samples_expected: string;
    mp_expected: string;
    for_valuation: string;
    req_disp_tech: string;
    req_disp_size: string;
    req_disp_res: string;
    req_disp_brigt: string;
    req_disp_angle: string;
    req_disp_od: string;
    req_disp_aa: string;
    req_disp_inter: string;
    req_disp_ot: string;
    req_disp_st: string;
    req_disp_spec: string;
    req_tp_size: string;
    req_tp_aa: string;
    req_tp_tech: string;
    req_tp_od: string;
    req_tp_inter: string;
    req_tp_glass: string;
    req_tp_spec: string;
    req_others: string;
  }) {
    try {
      const result = await pool.query(
        `INSERT INTO rfqs (
          rfq_code,
          clickup_id,
          sp,
          year,
          serial,
          department,
          eau,
          project_client_id,
          name,
          pm_id,
          samples_expected,
          mp_expected,
          for_valuation,
          req_disp_tech,
          req_disp_size,
          req_disp_res,
          req_disp_brigt,
          req_disp_angle,
          req_disp_od,
          req_disp_aa,
          req_disp_inter,
          req_disp_ot,
          req_disp_st,
          req_disp_spec,
          req_tp_size,
          req_tp_aa,
          req_tp_tech,
          req_tp_od,
          req_tp_inter,
          req_tp_glass,
          req_tp_spec,
          req_others)
          VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32)
          RETURNING id, rfq_code;`,
        [
          rfq_code,
          clickup_id,
          sp,
          year + "",
          serial + "",
          department,
          eau,
          project_client_id,
          name,
          pm_id,
          samples_expected,
          mp_expected,
          for_valuation,
          req_disp_tech,
          req_disp_size,
          req_disp_res,
          req_disp_brigt,
          req_disp_angle,
          req_disp_od,
          req_disp_aa,
          req_disp_inter,
          req_disp_ot,
          req_disp_st,
          req_disp_spec,
          req_tp_size,
          req_tp_aa,
          req_tp_tech,
          req_tp_od,
          req_tp_inter,
          req_tp_glass,
          req_tp_spec,
          req_others,
        ]
      );
      return result?.rows[0];
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async updateData({
    id,
    eau,
    pm_id,
    name,
    samples_expected,
    mp_expected,
    for_valuation,
    req_disp_tech,
    req_disp_size,
    req_disp_res,
    req_disp_brigt,
    req_disp_angle,
    req_disp_od,
    req_disp_aa,
    req_disp_inter,
    req_disp_ot,
    req_disp_st,
    req_disp_spec,
    req_tp_size,
    req_tp_aa,
    req_tp_tech,
    req_tp_od,
    req_tp_inter,
    req_tp_glass,
    req_tp_spec,
    req_others,
    final_solutions,
  }: {
    id: string;
    eau: string;
    pm_id: string;
    name: string;
    samples_expected: string;
    mp_expected: string;
    for_valuation: string;
    req_disp_tech: string;
    req_disp_size: string;
    req_disp_res: string;
    req_disp_brigt: string;
    req_disp_angle: string;
    req_disp_od: string;
    req_disp_aa: string;
    req_disp_inter: string;
    req_disp_ot: string;
    req_disp_st: string;
    req_disp_spec: string;
    req_tp_size: string;
    req_tp_aa: string;
    req_tp_tech: string;
    req_tp_od: string;
    req_tp_inter: string;
    req_tp_glass: string;
    req_tp_spec: string;
    req_others: string;
    final_solutions: string;
  }) {
    try {
      const result = await pool.query(
        `UPDATE rfqs SET
          eau = $2,
          pm_id = $3,
          name = $4,
          samples_expected = $5,
          mp_expected = $6,
          for_valuation = $7,
          req_disp_tech = $8,
          req_disp_size = $9,
          req_disp_res = $10,
          req_disp_brigt = $11,
          req_disp_angle = $12,
          req_disp_od = $13,
          req_disp_aa = $14,
          req_disp_inter = $15,
          req_disp_ot = $16,
          req_disp_st = $17,
          req_disp_spec = $18,
          req_tp_size = $19,
          req_tp_aa = $20,
          req_tp_tech = $21,
          req_tp_od = $22,
          req_tp_inter = $23,
          req_tp_glass = $24,
          req_tp_spec = $25,
          req_others = $26,
          final_solutions = $27,
          updated_at = CURRENT_TIMESTAMP
          WHERE id = $1
          RETURNING id, rfq_code;`,
        [
          id,
          eau,
          pm_id,
          name,
          samples_expected,
          mp_expected,
          for_valuation,
          req_disp_tech,
          req_disp_size,
          req_disp_res,
          req_disp_brigt,
          req_disp_angle,
          req_disp_od,
          req_disp_aa,
          req_disp_inter,
          req_disp_ot,
          req_disp_st,
          req_disp_spec,
          req_tp_size,
          req_tp_aa,
          req_tp_tech,
          req_tp_od,
          req_tp_inter,
          req_tp_glass,
          req_tp_spec,
          req_others,
          final_solutions,
        ]
      );
      return result?.rows[0];
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async delete(id: string) {
    if (id === "1") {
      throw new BadRequestError(
        `This is "SPECIAL" RFQ - you cannot delete it!`
      );
    }
    try {
      const result = await pool.query(
        `DELETE FROM rfqs WHERE id = $1 RETURNING id, rfq_code;`,
        [id]
      );
      return result?.rows[0];
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }

  static async count() {
    try {
      const result = await pool.query(`SELECT COUNT(*) FROM rfqs;`);
      return parseInt(result?.rows[0].count);
    } catch (error: any) {
      throw new BadRequestError(error.message);
    }
  }
}

export { RfqRepo };
